[clang][C++20 modules] textual `#include` sandwiched between two `import` declarations loses enum declarations

[bug.zip](https://github.com/user-attachments/files/26289488/bug.zip)

## Summary

When a textual `#include <limits>` appears between two `import` declarations of
named C++20 modules that both transitively include `<limits>` in their global
module fragments, Clang loses visibility of enumerators from `<limits>`
(`std::denorm_absent`, `std::round_toward_zero`).  Unqualified lookup for these
enumerators inside `namespace std {}` fails with "use of undeclared identifier".

Moving the `#include <limits>` before both imports, or after both imports, makes
the code compile.  The bug is triggered specifically by the sandwiched position.

## Reproducer

Four files are provided:

| File                     | Purpose                                           |
|--------------------------|---------------------------------------------------|
| `modA.cppm`              | Module whose GMF includes `<limits>` (via `<chrono>`) |
| `modB.cppm`              | Module whose GMF also includes `<limits>`          |
| `consumer.cc`            | **Fails** — `#include <limits>` between two imports |
| `consumer_workaround.cc` | **Works** — `#include <limits>` before both imports |
| `run.sh`                 | Builds modules and compiles both consumers         |

### Quick reproduction

```bash
chmod +x run.sh
./run.sh
```

### Manual reproduction

```bash
clang++ -std=gnu++23 -x c++-module --precompile -o modA.pcm modA.cppm
clang++ -std=gnu++23 -x c++-module --precompile -o modB.pcm modB.cppm

# This FAILS:
clang++ -std=gnu++23 \
    -fmodule-file=modA=modA.pcm \
    -fmodule-file=modB=modB.pcm \
    -fsyntax-only consumer.cc

# This succeeds:
clang++ -std=gnu++23 \
    -fmodule-file=modA=modA.pcm \
    -fmodule-file=modB=modB.pcm \
    -fsyntax-only consumer_workaround.cc
```

### Failing consumer (`consumer.cc`)

```cpp
import modA;          // GMF includes <limits>
#include <limits>     // textual include — between two imports
import modB;          // GMF also includes <limits>

struct MyInt {};
namespace std {
template <>
class numeric_limits<MyInt> {
 public:
  static constexpr float_denorm_style has_denorm = denorm_absent;       // ERROR
  static constexpr float_round_style round_style = round_toward_zero;   // ERROR
};
}
```

### Error output

```
consumer.cc:10:52: error: use of undeclared identifier 'denorm_absent'
   10 |   static constexpr float_denorm_style has_denorm = denorm_absent;
      |                                                    ^~~~~~~~~~~~~
consumer.cc:11:52: error: use of undeclared identifier 'round_toward_zero'
   11 |   static constexpr float_round_style round_style = round_toward_zero;
      |                                                    ^~~~~~~~~~~~~~~~~
```

## Analysis

The code is valid C++.  Inside `namespace std {}`, unqualified lookup for
`denorm_absent` must find the enumerator of `std::float_denorm_style` declared
in `<limits>`, which was textually included.

The bug appears to involve Clang's handling of header deduplication across module
boundaries:

1. `import modA;` is processed.  modA's BMI records that its GMF included
   `<limits>`.

2. `#include <limits>` is processed textually.  The declarations from `<limits>`
   (including `enum float_denorm_style { denorm_absent, ... }`) become visible
   in the current TU.

3. `import modB;` is processed.  modB's BMI also records that its GMF included
   `<limits>`.  At this point, Clang appears to incorrectly invalidate or
   shadow the textual declarations from step 2, presumably during some
   deduplication or merging pass that reconciles the three copies of `<limits>`
   (two from GMFs, one textual).

4. Subsequent unqualified lookup for `denorm_absent` inside `namespace std`
   fails because the enumerator declarations are no longer reachable.

When `#include <limits>` comes before both imports (the workaround), the textual
declarations are established first and the two import-time merges do not
invalidate them.

## Key observations from bisection

- The bug does NOT reproduce with just one module.  Two modules whose GMFs both
  include `<limits>` are required.
- The bug does NOT reproduce if the textual `#include <limits>` is before both
  imports or after both imports.  It must be sandwiched between them.
- The `#include <limits>` does not need to be literally in the `.cc` file — it
  can come from a transitively included header.  In our real-world case,
  `import fmt;` is at the top of `.cc` files, then a header transitively does
  `#include <limits>` followed by `import seastar;`.
- The enumerators affected are specifically from `enum float_denorm_style` and
  `enum float_round_style`.  The `numeric_limits` class template itself remains
  visible (the type names `float_denorm_style` and `float_round_style` are
  found; only their enumerators are lost).
- The specific header that gets "lost" does not need to be `<limits>`.  We
  suspect the bug generalizes to any standard library header that is textually
  included between two imports of modules that both have that header in their
  GMFs. We have not tested other headers.

## Environment

- Clang 21.1.6 (commit a832a5222e489298337fbb5876f8dcaf072c5cca)
- Also reproduced on Clang 21.1.8
- libstdc++ 15 (GCC 15.x)
- `-std=gnu++23`
- Fedora 43 x86_64

## Real-world impact

This bug blocks adopting C++20 named modules in projects that use multiple
modules with overlapping standard library headers in their GMFs — a very common
pattern.  In our case (ScyllaDB), `.cc` files begin with `import fmt;` and then
transitively include a header that does `import seastar;`.  Both the fmt and
seastar modules have `<limits>` in their GMFs (via transitive includes of
`<chrono>`, `<seastar/core/bitops.hh>`, etc.).  Any subsequent textual inclusion
of abseil headers that specialize `std::numeric_limits` fails.

The workaround is to ensure no textual `#include` of `<limits>` appears between
the two import declarations, which in practice means removing a defensive
`#include <limits>` that was placed (ironically) as a workaround for what
appeared to be a different manifestation of the same underlying issue.

--- a/clang/include/clang/AST/DeclBase.h
+++ b/clang/include/clang/AST/DeclBase.h
@@ -228,6 +228,14 @@ public:
     /// module is imported.
     VisibleWhenImported,
 
+    /// This declaration has an owning module, and is not visible to the
+    /// current TU but we promoted it to be visible for various reasons,
+    /// e.g., we have the same declaration in the current TU but we'd
+    /// like to avoid parsing it again.
+    ///
+    /// The vibility should be never be serialized.
+    VisiblePromoted,
+
     /// This declaration has an owning module, and is visible to lookups
     /// that occurs within that module. And it is reachable in other module
     /// when the owning module is transitively imported.
@@ -668,7 +676,7 @@ public:
   bool isInExportDeclContext() const;
 
   bool isInvisibleOutsideTheOwningModule() const {
-    return getModuleOwnershipKind() > ModuleOwnershipKind::VisibleWhenImported;
+    return getModuleOwnershipKind() > ModuleOwnershipKind::VisiblePromoted;
   }
 
   /// Whether this declaration comes from another module unit.
@@ -870,6 +878,11 @@ public:
   void setVisibleDespiteOwningModule() {
     if (!isUnconditionallyVisible())
       setModuleOwnershipKind(ModuleOwnershipKind::Visible);
+  }
+
+  void setVisiblePromoted() {
+    if (isInvisibleOutsideTheOwningModule() && isFromASTFile())
+      setModuleOwnershipKind(ModuleOwnershipKind::VisiblePromoted);
   }
 
   /// Get the kind of module ownership for this declaration.

--- a/clang/include/clang/AST/DeclContextInternals.h
+++ b/clang/include/clang/AST/DeclContextInternals.h
@@ -174,7 +174,9 @@ public:
     // Remove all declarations that are either external or are replaced with
     // external declarations with higher visibilities.
     DeclListNode::Decls *Tail = erase_if([Decls](NamedDecl *ND) {
-      if (ND->isFromASTFile())
+      // If the declaration is promoted intentionally, keep it.
+      if (ND->isFromASTFile() && ND->getModuleOwnershipKind() !=
+                                     Decl::ModuleOwnershipKind::VisiblePromoted)
         return true;
       // FIXME: Can we get rid of this loop completely?
       return llvm::any_of(Decls, [ND](NamedDecl *D) {

--- a/clang/lib/AST/Decl.cpp
+++ b/clang/lib/AST/Decl.cpp
@@ -1193,6 +1193,7 @@ static bool isExportedFromModuleInterfaceUnit(const NamedDecl *D) {
   case Decl::ModuleOwnershipKind::Unowned:
   case Decl::ModuleOwnershipKind::ReachableWhenImported:
   case Decl::ModuleOwnershipKind::ModulePrivate:
+  case Decl::ModuleOwnershipKind::VisiblePromoted:
     return false;
   case Decl::ModuleOwnershipKind::Visible:
   case Decl::ModuleOwnershipKind::VisibleWhenImported:

--- a/clang/lib/AST/DeclBase.cpp
+++ b/clang/lib/AST/DeclBase.cpp
@@ -402,7 +402,8 @@ void Decl::setLexicalDeclContext(DeclContext *DC) {
   }
 
   assert(
-      (getModuleOwnershipKind() != ModuleOwnershipKind::VisibleWhenImported ||
+      ((getModuleOwnershipKind() != ModuleOwnershipKind::VisibleWhenImported &&
+        getModuleOwnershipKind() != ModuleOwnershipKind::VisiblePromoted) ||
        getOwningModule()) &&
       "hidden declaration has no owning module");
 }

--- a/clang/lib/Sema/SemaLookup.cpp
+++ b/clang/lib/Sema/SemaLookup.cpp
@@ -1575,7 +1575,7 @@ void Sema::makeMergedDefinitionVisible(NamedDecl *ND) {
   if (auto *ED = dyn_cast<EnumDecl>(ND);
       ED && ED->isFromGlobalModule() && !ED->isScoped()) {
     for (auto *ECD : ED->enumerators()) {
-      ECD->setVisibleDespiteOwningModule();
+      ECD->setVisiblePromoted();
       DeclContext *RedeclCtx = ED->getDeclContext()->getRedeclContext();
       if (RedeclCtx->lookup(ECD->getDeclName()).empty())
         RedeclCtx->makeDeclVisibleInContext(ECD);

--- a/clang/lib/Serialization/ASTReaderDecl.cpp
+++ b/clang/lib/Serialization/ASTReaderDecl.cpp
@@ -636,6 +636,7 @@ void ASTDeclReader::VisitDecl(Decl *D) {
       break;
     case Decl::ModuleOwnershipKind::Unowned:
     case Decl::ModuleOwnershipKind::VisibleWhenImported:
+    case Decl::ModuleOwnershipKind::VisiblePromoted:
     case Decl::ModuleOwnershipKind::ReachableWhenImported:
     case Decl::ModuleOwnershipKind::ModulePrivate:
       break;

--- a/clang/test/Modules/include-between-imports-enums.cppm
+++ b/clang/test/Modules/include-between-imports-enums.cppm
@@ -1,0 +1,47 @@
+// RUN: rm -rf %t
+// RUN: mkdir -p %t
+// RUN: split-file %s %t
+//
+// RUN: %clang_cc1 -std=c++20 %t/A.cppm -emit-module-interface -o %t/A.pcm
+// RUN: %clang_cc1 -std=c++20 %t/B.cppm -emit-module-interface -o %t/B.pcm
+// RUN: %clang_cc1 -std=c++20 %t/use.cpp -fprebuilt-module-path=%t -verify -fsyntax-only
+//
+// RUN: %clang_cc1 -std=c++20 %t/A.cppm -emit-reduced-module-interface -o %t/A.pcm
+// RUN: %clang_cc1 -std=c++20 %t/B.cppm -emit-reduced-module-interface -o %t/B.pcm
+// RUN: %clang_cc1 -std=c++20 %t/use.cpp -fprebuilt-module-path=%t -verify -fsyntax-only
+
+// Test that a textual #include sandwiched between two import declarations
+// of modules that both include the same header in their GMFs does not lose
+// enum declarations. See https://github.com/llvm/llvm-project/issues/188853
+
+//--- enum.h
+#ifndef ENUM_H
+#define ENUM_H
+namespace ns {
+enum E { Value1, Value2, Value3 };
+}
+#endif
+
+//--- A.cppm
+module;
+#include "enum.h"
+export module A;
+export auto a = ns::Value1;
+
+//--- B.cppm
+module;
+#include "enum.h"
+export module B;
+export auto b = ns::Value2;
+
+//--- use.cpp
+// expected-no-diagnostics
+import A;
+#include "enum.h"
+import B;
+
+auto x = ns::Value3;
+
+namespace ns {
+auto y = Value1;
+}

